### PR TITLE
nixos-render-docs: init, use for some manual rendering to docbook

### PR DESCRIPTION
**Pull request by pennae**
###### Description of changes

another step toward removing docbook. what was once mergeJSON and is now optionsToDocbook is a good start towards more capable renderers, so let's extend that base into a new tool and use it for both rendering the options docbook fragment and the `manual.xml` fragment containing all the module chapters.

this can (and will) be extended further: we could sort the manual chapters by their title instead of file system path (which would greatly improve navigability of the module chapters fragment), we could include non-module chapters and release notes (obviating he need for md-to-db.sh), we could add more output formats (like html, man, at some point perhaps even info and epub).

the foundation laid here might even be altered slightly to allow for nixos-render-docs plugins that target entirely different formats, or maybe just collect weird and wonderful information about the documentation.

rendering changes from this PR are minimal and mostly confined to what smartquotes does to the options fragment, though a few docbook quotes in module chapters got turned into non-docbook quotes because pandoc is a bit extra like that.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/doc/manual/development/meta-attributes.section.md [hunk 1]
```diff
@@ -23,15 +23,15 @@ file.
...
+    doc = ./default.md;
     buildDocsInSandbox = true;
   };
 }
 ```
 
 -   `maintainers` contains a list of the module maintainers.
 
--   `doc` points to a valid DocBook file containing the module
+-   `doc` points to a valid CommonMark file containing the module
```
Comment (ncfavier) [inline nixos/doc/manual/development/meta-attributes.section.md:34]: Maybe we can point to https://nixos.org/manual/nixpkgs/unstable/#sec-contributing-markup here?

File: nixos/doc/manual/default.nix [hunk 2]
```diff
@@ -68,12 +68,15 @@ let
...
-    </section>
+  modulesDoc = runCommand "modules.xml" {
+    nativeBuildInputs = [ pkgs.nixos-render-docs ];
+  } ''
+    nixos-render-docs manual docbook \
+      --manpage-urls ${pkgs.path + "/doc/manpage-urls.json"} \
+      "$out" \
+      --section \
+        --section-id modules \
+        --chapters ${lib.concatMapStrings (p: "${p.value} ") config.meta.doc}
```
Comment (ncfavier) [inline nixos/doc/manual/default.nix:79]:
```suggestion
        --chapters ${lib.escapeShellArgs (map (p: p.value) config.meta.doc)}
```

File: pkgs/tools/nix/nixos-render-docs/default.nix [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    mdit-py-plugins-no-tests
+    python.pkgs.frozendict
+  ];
+
+  pytestFlagsArray = [ "-vvrP" "tests/" ];
+
+  meta = with lib; {
+    description = "Renderer for NixOS manual and option docs";
+    license = licenses.mit;
+    maintainers = [ ];
```
Comment (ncfavier) [inline pkgs/tools/nix/nixos-render-docs/default.nix:49]: You should probably add yourself :p

**Review by ncfavier — COMMENTED** [review]
This is fantastic.

File: pkgs/tools/nix/nixos-render-docs/default.nix (hunk 3 above)
Comment (ncfavier) [reply pkgs/tools/nix/nixos-render-docs/default.nix:49]: That's fair enough.

**Review by ncfavier — APPROVED** [review]

File: nixos/lib/make-options-doc/optionsToDocbook.py [hunk 4]
```diff
@@ -187,21 +186,31 @@ def myst_role(self, token, tokens, i, options, env):
...
-md = (
-    markdown_it.MarkdownIt(renderer_cls=Renderer)
-    # TODO maybe fork the plugin and have only a single rule for all?
-    .use(container_plugin, name="{.note}")
-    .use(container_plugin, name="{.important}")
-    .use(container_plugin, name="{.warning}")
-    .use(deflist_plugin)
-    .use(myst_role_plugin)
-)
+class Converter:
```
Comment (jtojnar) [inline nixos/lib/make-options-doc/optionsToDocbook.py]: Maybe use a bit more descriptive class name?

File: pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/__init__.py [hunk 5]
```diff
@@ -344,56 +344,56 @@ def print_decl_def(header, locs):
...
+    keys.sort(key=lambda opt: [ (0 if p.startswith("enable") else 1 if p.startswith("package") else 2, p)
+                                for p in options[opt]['loc'] ])
 
-print(f"""<?xml version="1.0" encoding="UTF-8"?>""")
-if OTD_DOCUMENT_TYPE == 'appendix':
-    print("""<appendix xmlns="http://docbook.org/ns/docbook" xml:id="appendix-configuration-options">""")
-    print("""  <title>Configuration Options</title>""")
-print(f"""<variablelist xmlns:xlink="http://www.w3.org/1999/xlink"
-                        xmlns:nixos="tag:nixos.org"
-                        xmlns="http://docbook.org/ns/docbook"
```
Comment (jtojnar) [inline pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/__init__.py:309]: This was lost accidentally?

File: pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/md.py [hunk 6]
```diff
@@ -10,3 +17,160 @@
...
+        def do_one(i: int, token: Token) -> str:
+            if token.type in self.rules:
+                return self.rules[token.type](tokens[i], tokens, i, options, env)
+            else:
+                raise NotImplementedError("md token not supported yet", token)
+        return "".join(map(lambda arg: do_one(*arg), enumerate(tokens)))
+
+    def text(self, token: Token, tokens: Sequence[Token], i: int, options: OptionsDict,
+             env: MutableMapping[str, Any]) -> str:
+        raise RuntimeError("md token not supported", token)
```
Comment (jtojnar) [inline pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/md.py:113]: Why not use use [`NotImplementedError`](https://docs.python.org/3/library/exceptions.html#NotImplementedError) here as well?

File: pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/__init__.py [hunk 7]
```diff
@@ -21,14 +21,16 @@
...
+
         self._md = markdown_it.MarkdownIt(
             "commonmark",
             {
                 'maxNesting': 100,   # default is 20
                 'html': False,       # not useful since we target many formats
                 'typographer': True, # required for smartquotes
             },
-            renderer_cls=DocBookRenderer
+            renderer_cls=lambda p: DocBookRenderer(self._manpage_urls, p)
```
Comment (jtojnar) [inline pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/__init__.py]: `parser` might be cleaner name, it is what the `RendererHTML` [uses in its constructor](https://github.com/executablebooks/markdown-it-py/blob/4e8947419b87f5d80fd46367e63ed2aa03307a46/markdown_it/renderer.py#L65).

```suggestion
            renderer_cls=lambda parser: DocBookRenderer(self._manpage_urls, parser)
```

File: pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/__init__.py [hunk 8]
```diff
@@ -47,16 +48,10 @@ def render(self, src: str) -> str:
 
 # converts in-place!
 def convertMD(options: Dict[str, Any]) -> str:
-    def optionIs(option: Dict[str, Any], key: str, typ: str) -> bool:
-        if key not in option: return False
-        if type(option[key]) != dict: return False
-        if '_type' not in option[key]: return False
-        return option[key]['_type'] == typ
-
     def convertCode(name: str, option: Dict[str, Any], key: str):
```
Comment (jtojnar) [inline pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/__init__.py]: Should this be an `Option` type too, when it is passed to `option_is`?

File: pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/md.py [hunk 9]
```diff
@@ -212,6 +215,51 @@ def dd_close(self, token: Token, tokens: Sequence[Token], i: int, options: Optio
...
+
+def _is_escaped(srcCharCode: tuple[int, ...], pos: int) -> bool:
+    found = 0
+    while pos >= 0 and srcCharCode[pos] == 92: # ord('\\')
+        found += 1
+        pos -= 1
+    return found % 2 == 1
+
+_INLINE_ANCHOR_PATTERN = re.compile(r"\{\s*#([\w-]+)\s*\}")
```
Comment (jtojnar) [inline pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/md.py:245]: Maybe also underscore? https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/id

File: pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/md.py [hunk 10]
```diff
@@ -212,6 +215,51 @@ def dd_close(self, token: Token, tokens: Sequence[Token], i: int, options: Optio
...
+
+def _inline_anchor_plugin(md: markdown_it.MarkdownIt) -> None:
+    def inline_anchor(state: markdown_it.rules_inline.StateInline, silent: bool) -> bool:
+        if state.srcCharCode[state.pos] != 91: # [
```
Comment (jtojnar) [inline pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/md.py:249]: Maybe define a constant like `LEFT_BRACE_ASCII_CODE = ord('[')`

File: nixos/lib/make-options-doc/default.nix [hunk 11]
```diff
@@ -148,42 +148,19 @@ in rec {
...
-          };
-          mdit-py-plugins = p.mdit-py-plugins.override {
-            inherit markdown-it-py;
-            disableTests = true;
-          };
-        in [
-          markdown-it-py
-          mdit-py-plugins
-        ]))
+      pkgs.nixos-render-docs
```
Comment (roberth) [inline nixos/lib/make-options-doc/default.nix:152]: This is fantastic!
It seems that `nixos/tests/installer.nix` can now be fixed in a somewhat durable manner.

File: pkgs/tools/nix/nixos-render-docs/default.nix [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    disableTests = true;
+  };
+in
+
+python.pkgs.buildPythonApplication {
+  pname = "nixos-render-docs";
+  version = "0.0";
+  format = "pyproject";
+
+  src = ./src;
```
Comment (roberth) [inline pkgs/tools/nix/nixos-render-docs/default.nix:31]: This being more of a project directory now, does it generate files that should be filtered and/or added to a local `.gitignore`?

File: nixos/doc/manual/development/meta-attributes.section.md [hunk 13]
```diff
@@ -23,15 +23,17 @@ file.
 
   meta = {
     maintainers = with lib.maintainers; [ ericsagnes ];
-    doc = ./default.xml;
```
Comment (roberth) [inline nixos/doc/manual/development/meta-attributes.section.md:26]: Looks like these generated xmls weren't checked in CI then? Good riddance. Although.. they may have diverged :confused:

File: pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/docbook.py [hunk 14]
```diff
@@ -25,16 +25,26 @@ def make_xml_id(s: str) -> str:
 class Deflist:
     has_dd = False
 
+Heading = NamedTuple('Heading', [('tag', str),
```
Comment (jtojnar) [inline pkgs/tools/nix/nixos-render-docs/src/nixos_render_docs/docbook.py:28]: Maybe something like `container_tag`?

**Review by jtojnar — APPROVED** [review]
LGTM.

**Pull request merged**